### DISCUSSION TITLE
Training EL pipeline - preprocess docs beforehand

### DISCUSSION THREAD

**Discussion by Joozty** [explosion/spaCy#7201] [Help: Coding & Implementations] [answered]
I was wondering whether preprocessing docs beforehand is ok.

Let me explain: I wanna train my model and try different parameters. The slowest part in the chain is [language call](https://github.com/explosion/projects/blob/v3/tutorials/nel_emerson/scripts/custom_functions.py#L21). I thought I could maybe preprocess docs beforehand in parallel so the training process would be faster. 

Is there something that I don't see an can cause issues?

The [function ](https://github.com/explosion/projects/blob/v3/tutorials/nel_emerson/scripts/custom_functions.py#L14-L21) would look like this:

```python
def read_files(file_gold_data: Path, file_entities: Path, nlp: "Language") -> Iterable[Example]:
    # docs with entity ruler run on it
    doc_bin_1 = DocBin().from_disk(file_entities)
    doc_bin_2 = DocBin().from_disk(file_gold_data)

    docs_1 = doc_bin_1.get_docs(nlp.vocab)
    docs_2 = doc_bin_2.get_docs(nlp.vocab)

    for doc_1, doc_2 in zip(docs_1, docs_2):
        yield Example(doc_1, doc_2)
```

**Comment by svlandeg** [comment, marked as answer]
Sure, I think that could work? Just make sure that your two arguments to the `Example` constructor are definitely aligned - i.e. refer to the same `doc.text` etc.

**Reply by Joozty** [reply]
Great. Thanks🙂 

~~I saw that examples are loaded every epoch. I wonder why? SpaCy should already have all information needed for training. Actually, that's why I was concerned to do it the way I explained above.~~
